src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -1341,6 +1343,96 @@ def is_ordinary(self, proof=True):
         """
         return not is_j_supersingular(self.j_invariant(), proof=proof)
 
+    def has_order(self, value, num_checks=8):
+        r"""
+        Return ``True`` if the curve has order ``value``.
+
+        INPUT:
+
+        - ``value`` -- integer in the Hasse-Weil range for this curve
+
+        - ``num_checks``-- integer (default: `8`); the number of times to check
+          whether ``value`` times a random point on this curve equals the
+          identity
+
+        .. NOTE::
+
+            Since the method is probabilistic, there is a possibility for the
+            method to yield false positives (i.e. returning ``True`` even when
+            the result is ``False``). Even worse, it is possible for this to
+            happen even when ``num_checks`` is increased arbitrarily. See below
+            for an example and :issue:`38617` for an open discussion.
+
+        EXAMPLES:
+
+        For curves over small finite fields, the order is computed and compared
+        directly::
+
+            sage: E = EllipticCurve(GF(7), [0, 1])
+            sage: E.order()
+            12
+            sage: E.has_order(12, num_checks=0)
+            True
+            sage: E.has_order(11, num_checks=0)
+            False
+            sage: E.has_order(13, num_checks=0)
+            False
+
+        This tests the method on a random curve::
+
+            sage: # long time (10s)
+            sage: p = random_prime(2**128, lbound=2**127)
+            sage: K = GF((p, 2), name="a")
+            sage: E = EllipticCurve(K, [K.random_element() for _ in range(2)])
+            sage: N = E.order()
+            sage: E.has_order(N, num_checks=20)
+            True
+            sage: E.has_order(N + 1)
+            False
+
+        This demonstrates the bug mentioned in the NOTE above. The return value
+        should be ``False`` after :issue:`38617` is fixed::
+
+            sage: E = EllipticCurve(GF(127), [0, 1])
+            sage: E.order()
+            108
+            sage: E.has_order(126)
+            True
+
+        AUTHORS:
+
+         - Mariah Lenox (2011-02-16): Initial implementation
+
+         - Gareth Ma (2024-01-21): Fix bug for small curves
+        """
+        # This method does *not* use the cached value of `_order` even when
+        # available.
+        q = self.base_field().order()
+        a, b = Hasse_bounds(q, 1)
+        if not a <= value <= b:
+            return False
+
+        # For really small values, the random tests are too weak to detect wrong
+        # orders So we go with computing directly instead.
+        if q <= 100:
```

Comment:
```suggestion
        if q <= 2^128:
```

`EllipticCurve([F.random_element(), F.random_element()]).order()` takes under a second (141 ms) for me for `F = GF(previous_prime(2^128))`, I think you can make the bound here much larger than 100. I'm not tied to `2^128` specifically, but 100 seems way too small.

For `F = GF(previous_prime(2^64))` it's pretty much instant (6.62 ms).

Reply:
6.62ms is quite far away from instant. Also, I am unable to reproduce your timings on my laptop, though admittedly it's a slow one.

```sage
sage: p = random_prime(2**64, lbound=2**63)
....: E = EllipticCurve(GF(p), j=randrange(p))
....: %time N = E.order(); N
....: %time E.has_order(N)
....: %time _ = E.random_point()
CPU times: user 44.7 ms, sys: 3.31 ms, total: 48 ms
Wall time: 55.8 ms
12084593079744292894
CPU times: user 5.8 ms, sys: 0 ns, total: 5.8 ms
Wall time: 5.87 ms
True
CPU times: user 214 μs, sys: 0 ns, total: 214 μs
Wall time: 217 μs```

Reply:
How about `q < 2^32` then? Or even `q < 1000`. I'd like if this was provably correct at least for the kinds of examples someone might do by hand.

Reply:
I will change it for now, but expect it to be reverted back once # 38617 is resolved.